OSDOCS-2427: Update best practices guidance

skopacz1 opened the pull request:
[OSDOCS-2427](https://issues.redhat.com/browse/OSDOCS-2427)

Version(s): 4.13+ (note to merge reviewer - the 4.13 auto CP will likely fail)

This PR adds guidance for users about best practices they can employ to ensure that their cluster update is successful. This is not meant to be a list of additional prerequisites that users must follow before initiating an update.

QE review:
- [x] QE has approved this change.

Preview: [Best practices for cluster updates](https://file.rdu.redhat.com/skopacz/OSDOCS-2427/updating/preparing_for_updates/updating-cluster-prepare.html#update-best-practices_updating-cluster-prepare) (VPN required)

wking commented on modules/update-best-practices.adoc:
```diff
@@ -0,0 +1,57 @@
+=== Ensure that the cluster is in an Upgradable state
+
+When one or more Operators have not reported their `Upgradeable` condition as `true` for more than an hour, the `ClusterNotUpgradeable` warning alert is triggered in the cluster.
+In most cases patch updates are not blocked by this alert, but you cannot perform a minor version update until this alert is resolved and all Operators report `Upgradeable` as `true`.
```
nit: `true` -> `True`  ([API][1]).

[1]: https://github.com/openshift/api/blob/03f577bfbd59e4058724de268e5669db7ee3865b/config/v1/types_cluster_operator.go#L110

petr-muller commented:
/cc

petr-muller commented on modules/update-best-practices.adoc:
```diff
@@ -0,0 +1,57 @@
+=== Choose versions recommended by the OpenShift Update Service
+
+The OpenShift Update Service (OSUS) provides update recommendations based on cluster characteristics such as the cluster's subscribed channel, which are then saved by the Cluster Version Operator as either recommended or conditional updates.
+While it is possible to attempt an update to a version that is not recommended by OSUS, doing so significantly increases the risk of update failure or unintended consequences to the cluster after the update has finished.
```
Can we spin this into a positive-sounding (or at least less negative) statement? I think "significantly increases risk" is not even that much true. "update failure" is very strong term, but actually very rare, most upgrade issues are annoyances. I'd say recommended paths are near-zero risk and non-recommended is low risk, we just don't have enough data to actually claim near-zero risk (even if it likely is safe - we do a *lot* of testing, and we do not have enough  resources to test every possible path).

So I would maybe just say that following recommended paths protects the users from encountering known issues or unintended consequences, or something like that.

skopacz1 replied:
I'll replace my wording with yours, sure thing.

petr-muller commented on modules/update-best-practices.adoc:
```diff
@@ -0,0 +1,57 @@
+
+[discrete]
+[id="nodes-ready_{context}"]
+=== Ensure that all nodes are available
```
I think "all nodes are available" is not exactly the condition that we want. The point is that the cluster should not be running with little to no spare capacity. If the cluster is "full", then taking down a node to upgrade it can be either slow  or even impossible. So I would describe this section as "make sure the cluster has enough spare capacity so that the upgrade process can actually take some nodes offline to update them. This has two flavors, (you already mention one of them):
1. There is spare capacity on compute nodes, so that when taking a node offline temporarily, its workloads can be placed on other nodes comfortably
2. Worker pools do not have enough unavailable notes to prevent taking more nodes offline temporarily.

LalatenduMohanty replied:
We have noticed that many times when a node is down MCO gets stuck. So I am supportive of the idea. I also agree that we should have spare nodes which would give some breathing room to pods.

petr-muller commented on modules/update-best-practices.adoc:
```diff
@@ -0,0 +1,57 @@
+=== Ensure that the cluster's PodDisruptionBudget is properly configured
+
+The `PodDisruptionBudget` object allows you to define the minimum number or percentage of pod replicas that must be available at any given time.
+This configuration allows workloads to be protected from disruptions during maintenance tasks such as cluster updates.
```
I think we can afford to be a bit more specific:

1. Make sure that for HA workloads, there are replicas that can be temporarily taken offline without PDB configuration prohibiting that
2. Make  sure that non-HA workloads are either not protected by PDBs, or that there is an alternative reliable mechanism that allows draining these workloads eventually (such as periodic restart)

LalatenduMohanty commented on modules/update-best-practices.adoc:
```diff
@@ -0,0 +1,57 @@
+Critical alerts must always be addressed as soon as possible, but it is especially important to address these alerts and resolve any problems before initiating a cluster update.
+Failing to address critical alerts before beginning an update can cause a loss of data or a major failure of cluster services.
+
+You should also periodically review Warning and Info alerts on the cluster to address any potentially problematic conditions before initiating an update.
```
I do not think we should ask customers to review info alerts , thats a lot of work without much value for update success. Just asking to review critical alerts is good enough.

skopacz1 replied:
Sounds good, I'll remove this sentence.

shellyyang1989 commented:
@JianLi-RH Could you please help take a look? Thank you!

JianLi-RH commented on modules/update-best-practices.adoc:
```diff
@@ -0,0 +1,63 @@
+
+[discrete]
+[id="recommended-versions_{context}"]
+=== Choose versions recommended by the OpenShift Update Service
```
I think it is better to let customers know how to get the OSUS recommended versions, like on web console or command line.

skopacz1 replied:
Unless I'm mistaken - I think as long as users follow our update procedures then they'll always use recommended versions, unless users look up how to find conditional updates or use `--to-image`. But those two options take more work to find in the docs.

So I think we already show users how to get recommended versions because that's the default method that we document in almost all of our procedures.

JianLi-RH replied:
um...you are right.

JianLi-RH commented on modules/update-best-practices.adoc:
```diff
@@ -0,0 +1,63 @@
+
+[discrete]
+[id="critical-alerts_{context}"]
+=== Address all critical alerts on the cluster
```
Where can customer find the critical alerts? and which alerts are critical? maybe a summarize here is helpful.

JianLi-RH replied:
I think run `oc adm upgrade` can also find the critical alerts. /cc @shellyyang1989 for confirm.

petr-muller replied:
Currently alerts are not surfaced in any `oc` command. OTA-1080 and OTA-1157 should eventually improve this state.

JianLi-RH commented on modules/update-best-practices.adoc:
```diff
@@ -0,0 +1,63 @@
+
+[discrete]
+[id="cluster-upgradeable_{context}"]
+=== Ensure that the cluster is in an Upgradable state
```
How do we know if a cluster is in an Upgradable state?

skopacz1 replied:
We document the Upgradeable condition of cluster Operators here: [Understanding cluster Operator condition types](https://docs.openshift.com/container-platform/4.14/updating/understanding_updates/intro-to-updates.html#understanding_clusteroperator_conditiontypes_understanding-openshift-updates).

I can add a link to that page underneath this overall best practices section

JianLi-RH replied:
That would be greatly help. Thanks!

skopacz1 commented on modules/update-best-practices.adoc:
```diff
@@ -0,0 +1,63 @@
+
+[discrete]
+[id="recommended-versions_{context}"]
+=== Choose versions recommended by the OpenShift Update Service
```
I can change it to be a second level heading, sure. I made it a third level for stylistic reasons but it's not super important to keep it that way. The reason they don't show up in the navigation on the right, however is because of the `[discrete]` elements, such as the one on line 14. 

I put those there because I thought it would overwhelm the sidebar navigation, but I guess if there are only 5 of these sections then that's okay for now!

JianLi-RH replied:
@skopacz1 I am sure you have updated the navigation level, but I can't see the change on preview page, could you help to confirm it?

skopacz1 replied:
The automated preview page is not working correctly, sorry about that. I've manually uploaded the most recent preview with your feedback here: https://file.rdu.redhat.com/skopacz/OSDOCS-2427/updating/preparing_for_updates/updating-cluster-prepare.html#update-best-practices_updating-cluster-prepare

(you'll need to be on VPN to access this site)

skopacz1 commented:
@JianLi-RH Sorry that I've taken so long to get back to your feedback! Would you mind taking another look and seeing if I addressed all of your points? Thanks!

skopacz1 commented:
/retest

skopacz1 commented:
Closing and reopening to restart automated tests and hopefully causing the preview to be updated

skopacz1 commented:
/retest

JianLi-RH commented:
Thank you @skopacz1.

/lgtm

skopacz1 commented:
/label peer-review-needed

bscott-rh commented on modules/update-best-practices.adoc:
```diff
@@ -0,0 +1,63 @@
+[id="update-best-practices_{context}"]
+= Best practices for cluster updates
+
+{product-title} is designed to provide a robust update experience that allows clusters to update with minimal disruptions to workloads.
```
```suggestion
{product-title} provides a robust update experience that minimizes workload disruptions.
```
Reworded to avoid passive voice and the use of "allow" per ISG:
"Avoid stating that inanimate objects grant abilities to people, as in “the product allows you to…”. Whenever possible, use a direct, user-focused alternative such as “you can use the product to…” or “with this product, users can…”.

bscott-rh commented on modules/update-best-practices.adoc:
```diff
@@ -0,0 +1,63 @@
+= Best practices for cluster updates
+
+{product-title} is designed to provide a robust update experience that allows clusters to update with minimal disruptions to workloads.
+Updates will not begin unless the cluster is determined to be in an upgradeable state at the time of the update request.
```
```suggestion
Updates will not begin unless the cluster is in an upgradeable state at the time of the update request.
```
Passive voice

bscott-rh commented on modules/update-best-practices.adoc:
```diff
@@ -0,0 +1,63 @@
+{product-title} is designed to provide a robust update experience that allows clusters to update with minimal disruptions to workloads.
+Updates will not begin unless the cluster is determined to be in an upgradeable state at the time of the update request.
+
+This design ensures that some key conditions are met before initiating an update, but there are a number of actions you can take to increase your chances of a successful cluster update.
```
```suggestion
This design enforces some key conditions before initiating an update, but there are a number of actions you can take to ensure a successful cluster update.
```
Reworded passive voice and added stronger tone. "increase your chances" sounds like the user is gambling

skopacz1 replied:
The first part of your suggestion sounds good, but I'm actively avoiding using terms like "ensuring" to avoid making promises to the user about the success of an update. The spirit of this doc is to acknowledge factors that can affect the success of an update and inform users how they can mitigate risk factors, it's not to assuage users that their updates will be universally successful.

IMO while we wouldn't want to explicitly state that updates are a gamble, that's sort of the nature of major tasks that introduce change and instability to the cluster, such as an update, and we don't want to misrepresent that nature in a doc meant to give users best practices and advice.

bscott-rh commented on modules/update-best-practices.adoc:
```diff
@@ -0,0 +1,63 @@
+//
+// * updating/preparing_for_updates/updating-cluster-prepare.adoc
+
+:_mod-docs-content-type: PROCEDURE
```
```suggestion
:_mod-docs-content-type: REFERENCE
```

skopacz1 replied:
Good catch, I copied and pasted this metadata stuff from another module

bscott-rh commented on modules/update-best-practices.adoc:
```diff
@@ -0,0 +1,63 @@
+[id="recommended-versions_{context}"]
+== Choose versions recommended by the OpenShift Update Service
+
+The OpenShift Update Service (OSUS) provides update recommendations based on cluster characteristics such as the cluster's subscribed channel, which are then saved by the Cluster Version Operator as either recommended or conditional updates.
```
```suggestion
The OpenShift Update Service (OSUS) provides update recommendations based on cluster characteristics such as the cluster's subscribed channel. The Cluster Version Operator saves these recommendations as either recommended or conditional updates.
```
Rewording suggestion

bscott-rh commented on modules/update-best-practices.adoc:
```diff
@@ -0,0 +1,63 @@
+Critical alerts must always be addressed as soon as possible, but it is especially important to address these alerts and resolve any problems before initiating a cluster update.
+Failing to address critical alerts before beginning an update can cause problematic conditions for the cluster.
+
+In the *Administrator* perspective of the web console, critical alerts can be
+found in *Observe -> Alerting*.
```
```suggestion
In the *Administrator* perspective of the web console, navigate to 
*Observe* -> *Alerting*.
```

skopacz1 replied:
If this suggestion is meant to remove passive voice, how about I rewrite it to this instead?
```suggestion
In the *Administrator* perspective of the web console, you can find critical alerts in *Observe -> Alerting*.
```

This way it still mentions critical alerts, which is why I'm telling users how to go to the alerting section of the UI.

bscott-rh commented on modules/update-best-practices.adoc:
```diff
@@ -0,0 +1,63 @@
+[id="cluster-upgradeable_{context}"]
+== Ensure that the cluster is in an Upgradable state
+
+When one or more Operators have not reported their `Upgradeable` condition as `True` for more than an hour, the `ClusterNotUpgradeable` warning alert is triggered in the cluster.
```
```suggestion
When one or more Operators have not reported their `Upgradeable` condition as `True` for more than an hour, the cluster triggers the `ClusterNotUpgradeable` warning alert.
```
passive voice

skopacz1 replied:
Sounds good, let me just check whether saying that it's the cluster that triggers the warning is accurate enough.

skopacz1 replied:
I've talked about this with an SME and the chain of events that cause this alert are complex and comprise multiple distinct components taking actions. I think I'll keep this passive, since the ISG states that the following two reasons, among others, are appropriate use cases for passive voice:
"Passive voice is acceptable when any of these conditions are true:
- The system performs the action.
- It is more appropriate to focus on the receiver of the action."

bscott-rh commented on modules/update-best-practices.adoc:
```diff
@@ -0,0 +1,63 @@
+== Ensure that the cluster is in an Upgradable state
+
+When one or more Operators have not reported their `Upgradeable` condition as `True` for more than an hour, the `ClusterNotUpgradeable` warning alert is triggered in the cluster.
+In most cases patch updates are not blocked by this alert, but you cannot perform a minor version update until this alert is resolved and all Operators report `Upgradeable` as `True`.
```
```suggestion
In most cases this alert does not block patch updates, but you cannot perform a minor version update until you resolve this alert and all Operators report `Upgradeable` as `True`.
```
passive voice

skopacz1 replied:
Sounds good for the first part, but as to changing the second part to "until you resolve this alert": I'm honestly not familiar enough with OCP to suggest that this alert always requires user intervention to resolve. Maybe in some cases something weird happens to an Operator for an hour to trigger the Upgradeable=false condition, then it fixes itself, reports `Upgradeable` as `true`, and the alert goes away solely due to system actions? 

I think I'll keep it as "this alert is resolved" to keep that user vs system ambiguity, unless an SME can confirm that basically every time an Operator reports `Upgradeable` as `false`, it's an issue that requires user action to resolve.

CC @petr-muller and @wking

skopacz1 replied:
Sounds good, I suppose the distinction is probably not relevant enough to leave that ambiguity in.

bscott-rh commented on modules/update-best-practices.adoc:
```diff
@@ -0,0 +1,63 @@
+== Ensure that enough spare nodes are available
+
+// Completely guessing the explanation in this section just to have something to start with when this is reviewed by an SME.
+A cluster should not be running with little to no spare node capacity, especially when initiating a cluster update.
+Nodes that are not running and available may limit a cluster's ability to perform an update with minimal disruption to cluster workloads.
```
I think it would be better to reword this in a positive voice, such as "You should have spare node capacity to ensure.." and "Having available nodes is important for cluster updates to complete with minimal disruption" or something to that effect. I'm not entirely sure what is meant by "running and available" though.

skopacz1 replied:
I'll admit this sounds negative but I think it's useful to talk about the consequences this can have on update success, especially since there's already a more positive sounding, actionable piece of advice at the end of the section here on line 49. I feel like it would be a bit redundant to also have similar phrasing here. That's just my opinion tho, let me know if you disagree.

bscott-rh commented on modules/update-best-practices.adoc:
```diff
@@ -0,0 +1,63 @@
+A cluster should not be running with little to no spare node capacity, especially when initiating a cluster update.
+Nodes that are not running and available may limit a cluster's ability to perform an update with minimal disruption to cluster workloads.
+
+Depending on the configured value of the cluster's `maxUnavailable` spec, an unavailable node can also prevent itself and other nodes from having machine configuration changes applied during a cluster update.
```
```suggestion
Depending on the configured value of the cluster's `maxUnavailable` spec, the cluster may not be able to apply machine configuration changes to nodes if there is an unavailable node.
```
passive voice

skopacz1 replied:
I can try and reword it in active voice, sure.

bscott-rh commented on modules/update-best-practices.adoc:
```diff
@@ -0,0 +1,63 @@
+[id="pod-disruption-budget_{context}"]
+== Ensure that the cluster's PodDisruptionBudget is properly configured
+
+The `PodDisruptionBudget` object allows you to define the minimum number or percentage of pod replicas that must be available at any given time.
```
```suggestion
You can use the `PodDisruptionBudget` object to define the minimum number or percentage of pod replicas that must be available at any given time.
```
Removed permissive language

bscott-rh commented on modules/update-best-practices.adoc:
```diff
@@ -0,0 +1,63 @@
+== Ensure that the cluster's PodDisruptionBudget is properly configured
+
+The `PodDisruptionBudget` object allows you to define the minimum number or percentage of pod replicas that must be available at any given time.
+This configuration allows workloads to be protected from disruptions during maintenance tasks such as cluster updates.
```
```suggestion
This configuration protects workloads from disruptions during maintenance tasks such as cluster updates.
```

bscott-rh reviewed (commented):
Nice work, left a few style guide comments.

/remove-label peer-review-in-progress
/remove-label peer-review-needed
/label peer-review-done

skopacz1 commented:
/label merge-review-needed

ShaunaDiaz commented:
/label merge-review-in-progress

ShaunaDiaz commented:
/remove-label merge-review-needed

ShaunaDiaz commented:
/cherrypick enterprise-4.15

ShaunaDiaz commented:
/cherrypick enterprise-4.14

ShaunaDiaz commented:
/cherrypick enterprise-4.13

ShaunaDiaz commented:
/remove-label merge-review-in-progress
